devdoc/channel.md

```diff
@@ -0,0 +1,297 @@
+
+    B --> N: channel_push
+    B --> R: channel_pull
+
+    N --> E:  channel_pull
+
+    R --> E: channel_push
+
+    N --> E:  cancel channel_push
+    R --> E: cancel channel_pull
+
+    R --> R: channel_pull
+    N --> N: channel_push
+
+    E --> [*]: channel_destroy
+```
+
+## Reentrancy
+
+`channel_pull` and `channel_push` can be called from callbacks of this module. Since the callbacks are executed on threadpool threads, there is no risk of stack overflow.
+
+## Internal Reference Counting
+
+`THANDLE(CHANNEL)` maintains a reference to an internal `THANDLE(CHANNEL_INTERNAL)`. `CHANNEL_INTERNAL` is the object that contains the state of the module. This two-level reference counting is needed because the lifetime of the module is linked to the `THANDLE(CHANNEL)` and the `THANDLE(ASYNC_OP)`s returned by `channel_pull` and `channel_push`. 
+
+## Exposed API
+```c
+#define CHANNEL_RESULT_VALUES \
+    CHANNEL_RESULT_OK, \
+    CHANNEL_RESULT_INVALID_ARGS, \
+    CHANNEL_RESULT_ERROR
+
+MU_DEFINE_ENUM(CHANNEL_RESULT, CHANNEL_RESULT_VALUES);
+
+#define CHANNEL_CALLBACK_RESULT_VALUES \
+    CHANNEL_CALLBACK_RESULT_OK, \
+    CHANNEL_CALLBACK_RESULT_CANCELLED, \
+    CHANNEL_CALLBACK_RESULT_ABANDONED
+
+MU_DEFINE_ENUM(CHANNEL_CALLBACK_RESULT, CHANNEL_CALLBACK_RESULT_VALUES);
+
+#include "umock_c/umock_c_prod.h"
+#ifdef __cplusplus
+extern "C" {
+#endif /* __cplusplus */
+
+typedef void(*PULL_CALLBACK)(void* pull_context, THANDLE(RC_PTR) data, CHANNEL_CALLBACK_RESULT result);
+typedef void(*PUSH_CALLBACK)(void* push_context, CHANNEL_CALLBACK_RESULT result);
+typedef struct CHANNEL_TAG CHANNEL;
+
+THANDLE_TYPE_DECLARE(CHANNEL);
+
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create, THANDLE(THREADPOOL), threadpool);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_pull, THANDLE(CHANNEL), channel, PULL_CALLBACK, pull_callback, void*, pull_context, THANDLE(ASYNC_OP)*, out_op_pull);
+    MOCKABLE_FUNCTION(, CHANNEL_RESULT, channel_push, THANDLE(CHANNEL), channel, THANDLE(RC_PTR), data, PUSH_CALLBACK, push_callback, void*, push_context, THANDLE(ASYNC_OP)*, out_op_push);
+
+#ifdef __cplusplus
+}
+#endif /* __cplusplus */
+
+```
+
+### channel_create
+```c
+    MOCKABLE_FUNCTION(, THANDLE(CHANNEL), channel_create, THANDLE(THREADPOOL), threadpool);
+```
+
+`channel_create` creates the channel and returns it.
+
+**SRS_CHANNEL_43_077: [** If `threadpool` is `NULL`, `channel_create` shall fail and return `NULL`. **]**
+
+**SRS_CHANNEL_43_001: [** `channel_create` shall create a `CHANNEL` object by calling `THANDLE_MALLOC` with `channel_dispose` as `dispose`. **]**
+
+**SRS_CHANNEL_43_078: [** `channel_create` shall create a `CHANNEL_INTERNAL` object by calling `THANDLE_MALLOC`.**]**
+
+**SRS_CHANNEL_43_079: [** `channel_create` shall store the created `THANDLE(CHANNEL_INTERNAL)` in the `THANDLE(CHANNEL)`. **]**
+
+**SRS_CHANNEL_43_080: [** `channel_create` shall initialize the `CHANNEL_INTERNAL` with the given `threadpool`. **]**
```

Comment:
>CHANNEL_INTERNAL [](http://example.com/codeflow?start=65&length=16)

I've looked everywhere and this CHANNEL_INTERNAL seems to be holding `threadpool`. But threadpool is itself a THANDLE.

It is really weird to have a THANDLE on top of a THANDLE.

Reply:
The section on Internal Reference Counting explains this.